src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,432 @@
+            Entry entry = request.getCacheEntry();
+            if (entry == null) {
+                callback.onSuccess(
+                        new NetworkResponse(
+                                HttpURLConnection.HTTP_NOT_MODIFIED,
+                                /* data= */ null,
+                                /* notModified= */ true,
+                                SystemClock.elapsedRealtime() - requestStartMs,
+                                responseHeaders));
+                return;
+            }
+            // Combine cached and response headers so the response will be complete.
+            List<Header> combinedHeaders = NetworkUtility.combineHeaders(responseHeaders, entry);
+            callback.onSuccess(
+                    new NetworkResponse(
+                            HttpURLConnection.HTTP_NOT_MODIFIED,
+                            entry.data,
+                            /* notModified= */ true,
+                            SystemClock.elapsedRealtime() - requestStartMs,
+                            combinedHeaders));
+            return;
+        }
+
+        byte[] responseContents = httpResponse.getContentBytes();
+
+        if (responseContents == null) {
+            final InputStream inputStream = httpResponse.getContent();
+            if (inputStream == null) {
+                // Add 0 byte response as a way of honestly representing a
+                // no-content request.
+                responseContents = new byte[0];
+            } else {
+                Runnable run =
+                        new Runnable() {
+                            @Override
+                            public void run() {
+                                PoolingByteArrayOutputStream bytes =
+                                        new PoolingByteArrayOutputStream(
+                                                mPool, httpResponse.getContentLength());
+                                byte[] buffer = new byte[1024];
+                                int count = 0;
+                                while (true) {
+                                    try {
+                                        if ((count = inputStream.read(buffer)) == -1) break;
+                                    } catch (IOException e) {
+                                        onRequestFailed(
+                                                request,
+                                                callback,
+                                                e,
+                                                requestStartMs,
+                                                httpResponse,
+                                                bytes.toByteArray());
+                                    }
+                                    bytes.write(buffer, 0, count);
+                                }
+                                byte[] finalResponseContents = bytes.toByteArray();
+                                runAfterBytesReceived(
```

Comment:
nit: perhaps "onResponseRead", which would be consistent with "onRequestSucceeded" or "onRequestFailed" naming-wise.